### PR TITLE
My ArrayList

### PR DESCRIPTION
**Pull request by CherednichenkoNikita**

### CODE REVIEW

File: src/main/java/com/java_professional/cherednichenko/myArrayList/Main.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        MyArrayList<String> strings = new MyArrayList<>();
+        strings.add("dddd");
+        strings.add("dddd");
+        strings.add("dddd");
+        strings.add("dddd");
+        strings.add("ddd");
+        strings.remove("dddd");
+        for (String s : strings) {
+            System.out.println(s);
+        }
```
Comment (ochedenis) [inline src/main/java/com/java_professional/cherednichenko/myArrayList/Main.java:15]: For print you list at console just realize method toString.

File: src/main/java/com/java_professional/cherednichenko/myArrayList/MyAbstractList.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
+package com.java_professional.cherednichenko.myArrayList;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+
+public abstract class MyAbstractList<T> implements Iterable<T> {
```
Comment (ochedenis) [inline src/main/java/com/java_professional/cherednichenko/myArrayList/MyAbstractList.java]: interface MyAbstractList<T> extends Iterable<T>

File: src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
+package com.java_professional.cherednichenko.myArrayList;
+
+import java.util.*;
+import java.util.function.Consumer;
+
+
+public class MyArrayList<T> extends MyAbstractList<T> {
+    private final int INIT_SIZE = 16;
+    private Object[] array = new Object[INIT_SIZE];
```
Comment (ochedenis) [inline src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java]: private T[] array = (T[]) new Object[INIT_SIZE];

File: src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    public T set(int index, T element) {
+        T oldValue = (T) array[index];
+        array[index] = element;
+        return oldValue;
+    }
+
+    public void resize(int newLength) {
+        Object[] newArray = new Object[newLength];
```
Comment (ochedenis) [inline src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java]: T[] newArray = (T[]) new Object[newLength];

File: src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        System.arraycopy(array, 0, newArray, 0, pointer);
+        array = newArray;
+    }
+
+    public void add(T element) {
+        if (pointer == array.length - 1)
+            resize(array.length * 2);
```
Comment (ochedenis) [inline src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java:67]: java code conventions
    if (pointer == array.length - 1) {
            resize(array.length * 2);
     }

File: src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    static <T> T elementAt(Object[] es, int index) {
+        return (T) es[index];
+    }
+
+    @Override
+    public void forEach(Consumer<? super T> action) {
+        Object[] objects = new Object[pointer];
+        System.arraycopy(array, 0, objects, 0, pointer);
+        Objects.requireNonNull(action);
+        for (int i = 0; i < pointer; i++)
```
Comment (ochedenis) [inline src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java]: java code conventions:
        for (int i = 0; i < pointer; i++) {
              action.accept(elementAt(objects, i));
        }

File: src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        int result = Objects.hash(INIT_SIZE, pointer);
+        result = 31 * result + Arrays.hashCode(array);
+        return result;
+    }
+
+    public T remove(T element) {
+        if (!contains(element)) return null;
+        if (pointer + 1 - indexOf(element) >= 0)
+            System.arraycopy(array, indexOf(element) + 1, array, indexOf(element),
+                    pointer + 1 - indexOf(element));
```
Comment (ochedenis) [inline src/main/java/com/java_professional/cherednichenko/myArrayList/MyArrayList.java:97]: Make "if" like it requires java code conventions.

**Review by ochedenis — CHANGES_REQUESTED** [review]

**Review by yuriy-tsarenko — APPROVED** [review]

**Pull request merged**